data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ParserTest.java

```diff
@@ -137,6 +137,21 @@ void testValidOptionalSpaceOperators(final String expression) {
         assertThatIsValid(expression);
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = {
+            "not false",
+            "not not false",
+            "not /status_code",
+            "not (true)",
+            "-5",
+            "-3.14",
+            "-(5)",
+            "--10",
+    })
```

Comment:
These expressions are also valid and pass this test. 
```
"- not 3.14"
"- not /status_code",
"- not - not - not - 6"
```

Reply:
I think it is almost impossible for parser to complain about the above cases unless there is some tricky workaround in the grammar definition. Right now it is evaluator's job to error out those.

Reply:
Do we want those to be valid?

Reply:
Enforcing static type checking based on the grammar definition is possible. A significant amount of effort will be required to add static type checking support. I'll open a github issue to track this as a feature request.